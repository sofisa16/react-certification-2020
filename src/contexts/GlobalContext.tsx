import React, {useState, useEffect, useReducer} from 'react';
import {ThemeProvider} from 'styled-components';
import NoSsr from '@material-ui/core/NoSsr';
import {
  createMuiTheme,
  ThemeProvider as MuiThemeProvider,
  Theme,
} from '@material-ui/core/styles';
import {initialState, reducer, FavoriteVideos, Action} from './../components/FavoritesView/addToFavorites';
import {AUTH_STORAGE_KEY, AUTH_AVATAR, FAVORITES_VIDEOS} from './../utils/constants';
import {storage} from './../utils/storage';
import { useAuth0 } from "@auth0/auth0-react";

interface GlobalContextValues {
  search: string;
  setSearch: React.Dispatch<React.SetStateAction<string>>;
  darkState: boolean;
  setDarkState: React.Dispatch<React.SetStateAction<boolean>>;
  theme: Theme;
  favoriteVideos: FavoriteVideos;
  dispatchFav: React.Dispatch<Action>;
  authenticated: boolean;
  setAuthenticated: React.Dispatch<React.SetStateAction<boolean>>;
  avatar: string;
  setAvatar: React.Dispatch<React.SetStateAction<string>>;
}

interface GlobalContextProviderProps {
  children: React.ReactElement | React.ReactElement[];
}

const GlobalContext = React.createContext<GlobalContextValues>({
  search: '',
  setSearch: () => { return; },
  darkState: false,
  setDarkState: () => { return; },
  theme: createMuiTheme({
    palette: {
      type: 'light',
    }
  }),
  favoriteVideos: {},
  dispatchFav: () => { return; },
  authenticated: false,
  setAuthenticated: () => { return; },
  avatar: '',
  setAvatar: () => { return; },
});

const GlobalContextProvider: React.FunctionComponent<GlobalContextProviderProps> =
  (props: GlobalContextProviderProps) => {
    const [search, setSearch] = useState<string>('wizeline');
    const [darkState, setDarkState] = useState<boolean>(false);
    const [theme, setTheme] = useState<Theme>(
      createMuiTheme({
        palette: {
          type: 'light',
        },
      })
    );
    const [favoriteVideos, dispatchFav] = useReducer(reducer, initialState);
    const [authenticated, setAuthenticated] = useState<boolean>(false);
    const [avatar, setAvatar] = useState<string>('');
    const { user, isAuthenticated } = useAuth0();

    useEffect(
      () => {
        setAuthenticated(isAuthenticated);
      },
      [isAuthenticated]
    );

    useEffect(
      () => {
        if (user) {
          setAvatar(user.picture);
        }
      },
      [user]
    );

    useEffect(
      () => {
        setAuthenticated(Boolean(storage.get(AUTH_STORAGE_KEY)));
        setAvatar(String(storage.get(AUTH_AVATAR)));
        dispatchFav({
          type: 'addAll',
          all: JSON.parse(String(storage.get(FAVORITES_VIDEOS))),
        });
      },
      []
    );

    useEffect(
      () => {
        const palletType = darkState ? 'dark' : 'light';
        setTheme(
          createMuiTheme({
            palette: {
              type: palletType,
            }
          })
        );
      },
      [darkState]
    );

    useEffect(
      () => {
        storage.set(FAVORITES_VIDEOS, JSON.stringify(favoriteVideos));
      },
      [favoriteVideos]
    );

    const value: GlobalContextValues = {
      search,
      setSearch,
      darkState,
      setDarkState,
      theme,
      favoriteVideos,
      dispatchFav,
      authenticated,
      setAuthenticated,
      avatar,
      setAvatar,
    };

    return (
      <NoSsr>
        <MuiThemeProvider theme={theme}>
          <ThemeProvider theme={theme}>
            <GlobalContext.Provider
              value={value}
            >
              {props.children}
            </GlobalContext.Provider>
          </ThemeProvider>
        </MuiThemeProvider>
      </NoSsr>
    );
  };

export {GlobalContextProvider, GlobalContext};