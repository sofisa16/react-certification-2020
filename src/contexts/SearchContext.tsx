import React, {useState} from 'react';

interface SearchContextValues {
  search: string;
  setSearch: React.Dispatch<React.SetStateAction<string>>;
}

interface SearchContextProviderProps {
  children: React.ReactElement | React.ReactElement[];
}

const SearchContext = React.createContext<SearchContextValues>({
  search: "",
  setSearch: () => { return; },
});

const SearchContextProvider: React.FunctionComponent<SearchContextProviderProps> =
  (props: SearchContextProviderProps) => {
    const [search, setSearch] = useState<string>('wizeline');
    const value: SearchContextValues = {
      search,
      setSearch,
    };

    return (
      <SearchContext.Provider
        value={value}
      >
        {props.children}
      </SearchContext.Provider>
    );
  };

export {SearchContextProvider, SearchContext};