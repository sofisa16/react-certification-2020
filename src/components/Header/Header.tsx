import React, {useContext, useState} from 'react';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import IconButton from '@material-ui/core/IconButton';
import MenuIcon from '@material-ui/icons/Menu';
import AccountCircle from '@material-ui/icons/AccountCircle';
import styled from 'styled-components';
import SearchBox from './components/SearchBox/SearchBox';
import Switch from '@material-ui/core/Switch';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import Drawer from '@material-ui/core/Drawer';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import {GlobalContext} from './../../contexts/GlobalContext';
import HomeIcon from '@material-ui/icons/Home';
import FavoriteIcon from '@material-ui/icons/Favorite';
import {useHistory, useLocation} from 'react-router-dom';
import Avatar from '@material-ui/core/Avatar';
import Menu from '@material-ui/core/Menu';
import MenuItem from '@material-ui/core/MenuItem';
import {AUTH_STORAGE_KEY, AUTH_AVATAR, FAVORITES_VIDEOS} from './../../utils/constants';
import {storage} from './../../utils/storage';
import { useAuth0 } from "@auth0/auth0-react";

const MenuMargin = styled.div`
  margin-right: 16px;
`;

const RightSide = styled.div`
  @media (min-width: 600px) {
    display: flex;
  }

  display: none;
  justify-content: flex-end;
  align-items: center;
`;

const ToolbarGrid = styled(Toolbar)`
  display: grid;
  grid-template-columns: auto auto 1fr;
`;

interface StyledAppBarProps {
  darkState: boolean;
}

const StyledAppBar = styled(AppBar)<StyledAppBarProps>`
  ${({ theme, darkState }) => `
    background-color: ${darkState ? '#1C5476' : theme.palette.primary.main}
  `}
`;

const StyledListItemIcon = styled(ListItemIcon)`
  min-width: 24px;
  padding-right: 12px;
`;

const StyledList = styled(List)`
  width: 250px;
`;

function Header(): JSX.Element {
  const { darkState, setDarkState, authenticated, avatar, setAuthenticated, setAvatar, dispatchFav } = useContext(GlobalContext);
  const [openDrawer, setOpenDrawer] = useState<boolean>(false);
  const history = useHistory();
  const { loginWithRedirect, logout } = useAuth0();
  const location = useLocation();

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    event.preventDefault();
    setDarkState(event.target.checked);
  };

  function onClick(): void {
    setOpenDrawer(false);
    history.push('/');
  }

  function onClickFav(): void {
    setOpenDrawer(false);
    history.push('/favorites');
  }

  const [anchorEl, setAnchorEl] = React.useState<null | HTMLElement>(null);

  const handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
    setAnchorEl(event.currentTarget);
  };

  const handleClose = () => {
    setAnchorEl(null);
  };

  function doOpenLogin(): void {
    history.push('/login', { background: location });
    handleClose();
  }

  function logOut(): void {
    setAuthenticated(false);
    setAvatar('');
    dispatchFav({
      type: 'removeAll',
    });
    storage.set(AUTH_STORAGE_KEY, false);
    storage.set(AUTH_AVATAR, '');
    storage.set(FAVORITES_VIDEOS, {});
    handleClose();
    logout({
      returnTo: 'https://sofisa16.netlify.app/',
    });
  }

  return (
    <div>
      <Drawer anchor='left' open={openDrawer} onClose={() => { setOpenDrawer(false); }}>
        <StyledList>
          <ListItem button key={'home'} onClick={onClick}>
            <StyledListItemIcon><HomeIcon /></StyledListItemIcon>
            <ListItemText primary={"Home"} />
          </ListItem>
          {
            authenticated &&
            <ListItem button key={'favorite'} onClick={onClickFav}>
              <StyledListItemIcon><FavoriteIcon /></StyledListItemIcon>
              <ListItemText primary={"Favorites"} />
            </ListItem>
          }
        </StyledList>
      </Drawer>
      <StyledAppBar position='static' darkState={darkState}>
        <ToolbarGrid>
          <MenuMargin>
            <IconButton
              edge='start'
              color='inherit'
              aria-label='open panel'
            >
              <MenuIcon onClick={() => { setOpenDrawer(true); }}/>
            </IconButton>
          </MenuMargin>
          <SearchBox />
          <RightSide>
            <FormControlLabel
              control={
                <Switch
                  checked={darkState}
                  onChange={handleChange}
                  name='darkTheme'
                  id='darkTheme'
                />
              }
              label='Dark mode'
            />
            <IconButton
              edge='end'
              aria-label='account of current user'
              aria-haspopup='true'
              color='inherit'
              onClick={handleClick}
            >
              {
                authenticated
                  ? <Avatar alt={'avatar'} src={avatar} />
                  : <AccountCircle />
              }
            </IconButton>
            <Menu
              id="login-menu"
              anchorEl={anchorEl}
              keepMounted
              open={Boolean(anchorEl)}
              onClose={handleClose}
            >
              {
                authenticated
                  ? <MenuItem onClick={logOut}>Log out</MenuItem>
                  : [
                      <MenuItem onClick={doOpenLogin} key='doOpenLogin'>
                        Login
                      </MenuItem>,
                      <MenuItem onClick={loginWithRedirect} key='loginWithRedirect'>Auth0 log in</MenuItem>
                    ]
              }
            </Menu>
          </RightSide>
        </ToolbarGrid>
      </StyledAppBar>
    </div>
  );
}

export default Header;