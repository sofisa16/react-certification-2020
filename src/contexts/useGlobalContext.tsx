import {useContext} from 'react';
import {GlobalContext, GlobalContextValues} from './GlobalContext';

function useGlobalContext(): GlobalContextValues {
  const context = useContext(GlobalContext);

  if (!context) {
    throw new Error('Please use the "useContext" hook with a GlobalContextProvider.')
  }

  return context;
}

export default useGlobalContext;