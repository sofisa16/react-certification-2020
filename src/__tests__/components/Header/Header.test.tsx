import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import Header from '../../../components/Header/Header';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';
import renderer from 'react-test-renderer';
import {GlobalContextProvider} from './../../../contexts/GlobalContext';
import CssBaseline from '@material-ui/core/CssBaseline';

test('renders Dark mode text', () => {
  const history = createMemoryHistory();
  const route = '/';
  history.push(route);

  render(
    <GlobalContextProvider>
      <CssBaseline>
        <Router history={history}>
          <Header />
        </Router>
      </CssBaseline>
    </GlobalContextProvider>
  );
  const text = screen.getByText(/Dark mode/i);
  expect(text).toBeInTheDocument();
});

test('toogle Dark mode', () => {
  const history = createMemoryHistory();
  const route = '/';
  history.push(route);

  const utils = render(
    <GlobalContextProvider>
      <CssBaseline>
        <Router history={history}>
          <Header />
        </Router>
      </CssBaseline>
    </GlobalContextProvider>
  );
  const input = utils.getByLabelText('Dark mode') as HTMLInputElement;
  fireEvent.change(input, { target: { checked: true } });
  expect(input).toBeChecked();
});

it('renders Header component', () => {
  const history = createMemoryHistory();
  const route = '/';
  history.push(route);

  const tree = renderer
    .create(
      <GlobalContextProvider>
        <CssBaseline>
          <Router history={history}>
            <Header />
          </Router>
        </CssBaseline>
      </GlobalContextProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});