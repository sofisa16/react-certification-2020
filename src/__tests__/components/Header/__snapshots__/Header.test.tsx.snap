// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Header component renders Header component 1`] = `null`;
